Added second challenge for Director of Product (#162)
gravitational/careers: challenges/product/challenge-2.md

# Summary

Interview process for a Director of Product, Data Platform that selects for the
skills needed to build data pipelines and a data warehouse.

## Rationale

This exercise has two goals:

It helps us understand what to expect from you as a technical leader and
whether you have the technical aptitude to engage with highly technical
customers. It gives you a feel for what it's like to work at Teleport. The
exercise simulates day-to-day work, helping you better understand the team and
our hiring process.

We believe this approach is not only more effective, but also more enjoyable
than the whiteboard or quiz-style interviews common in the industry.

We appreciate your time and look forward to hacking on this little project
together.

## Objective

> Design and build a data warehouse for a fictional business.

Reviewer: It's unclear what is the success criteria of the data warehouse as defined by us? What capabilities it should have?

You will receive raw data about the business and must build a data pipeline to process
and import the data into a data warehouse. Then you'll use the data to

> write a 1-page document that provides insight and suggests changes to that

Reviewer: What types of insights in the business, just generic or there is a specific angle?

fictional business.

## Requirements

* Determine implementation, quality, security, and scope requirements to write
  a technical design document
* Build data pipelines and data warehouse using the sample data we provide to you
* Write a business document that outlines the insights and changes to the

>   fictional business you are proposing

Reviewer 2: Where is the sample data and what's the business?

> * Explain the business and technical value of the proposed solution on a short

Reviewer 2: Do we want to define how end-users will access the data warehouse? e.g. a OSS GUI wrapper, CLI, or pre-canned reports?

  call

Write your data pipeline scripts in Go or Python. Use DuckDB as your data

> warehouse.

Reviewer 2: Any specific reason for DuckDB for this challenge? There could a be range of DB options that may solve the problem in another way... depending on what insights are needed.

voting system: +1 is submitted whenever a team member accepts the submission, -2
otherwise.

In case of a positive result, we will connect you to our HR team who will
collect one-two references and will work out other details. You can start the
reference collection process in parallel if you would like to speed up the
process.

After reference collection, our recruiting team will send you an offer.

In case of a negative score result, the hiring manager will contact you and
share a list of the key observations from the team that affected the result.

### Code and project ownership

This is a test challenge and we have no intent of using the design you've
submitted in production. This is your work, and you are free to do whatever you
feel is reasonable with it. In the scenario when you don't pass, you can open
source it with any license and use it as a portfolio project.

### Areas of focus

Teleport focuses on networking, infrastructure and security.

* Reproducible pipelines. Scripts should be written in a way to allow
  reproduction of environment.
* Data models. Design performant tables. Keep queries fast and plan for schema
  migration.

> * Data quality. Add validation and test coverage to all stages of your pipeline

Reviewer 2: re: Above comment on Sample Data. We should add both classic data quality problems, different date types, incorrect values to newer issues including but not limited to prompt injections, encoding artifact and structural injections

# Summary

Interview process for a Director of Product, Data Platform that selects for the
skills needed to build data pipelines and a data warehouse.

## Rationale

This exercise has two goals:

> It helps us understand what to expect from you as a technical leader, and

Reviewer 3: ```suggestion
It helps us understand what to expect from you as a technical leader and
```

whether you have the technical aptitude to engage with highly technical
customers. It gives you a feel for what it's like to work at Teleport. The
exercise simulates day-to-day work, helping you better understand the team and
our hiring process.

We believe this approach is not only more effective, but also more enjoyable
than the whiteboard or quiz-style interviews common in the industry.

We appreciate your time and look forward to hacking on this little project
together.

## Objective

Design and build a data warehouse for a fictional business.

You will receive raw data about the business, build a data pipeline to process
and import that data into a data warehouse, and then use that information to
write a 1-page document that provides insight and suggests changes to that

> fictional business.

Reviewer 3: ```suggestion
You will receive raw data about the business and must build a data pipeline to process
and import the data into a data warehouse. Then you'll use the data to
write a 1-page document that provides insight and suggests changes to that
fictional business.
```

## Requirements

* Determine implementation, quality, security, and scope requirements to write
  a technical design document
* Build data pipelines and data warehouse using the sample data we provide to you
* Write a business document that outlines the insights and changes to the
  fictional business you are proposing
* Explain the business and technical value of the proposed solution on a short
  call

Write your data pipeline scripts in Go or Python. Use DuckDB as your data
warehouse.

## Guidance

### Interview Process

You will join a Slack channel with the interview team. The team consists of the

> peers who will be working with you.

Reviewer 3: ```suggestion
You will join a Slack channel with the interview team that consists of
peers who will be working with you.
```

Ask them about the culture, work and life balance, or anything else that you
would like to learn about Teleport.

Before writing the actual code, create a brief design document in Markdown and

> share with the team via a GitHub Pull Request.

Reviewer 3: First mention of a GitHub pull request, before mentioning they should create a repository and add the interview team as collaborators.

There is no required format or word count for the design document. The design
should convey your plan to satisfy the requirements, possible alternatives, and
the tradeoffs informing your approach. The team will review your design and
provide feedback by commenting on the pull request.

After the team has approved your design document, you may begin submitting pull
requests with the implementation.

After the implementation is complete, prepare a presentation and demo and
schedule a 45 minute call with the interview team. During this call you will

> present your solution to the audience and answer any questions.

Reviewer 3: ```suggestion
After the implementation is complete, prepare a presentation and demo, then
schedule a 45 minute call with the interview team. During this call you will
present your solution to the interview team and answer any questions.
```

collect one-two references and will work out other details. You can start the
reference collection process in parallel if you would like to speed up the
process.

After reference collection, our recruiting team will send you an offer.

In case of a negative score result, the hiring manager will contact you and
share a list of the key observations from the team that affected the result.

### Code and project ownership

This is a test challenge and we have no intent of using the design you've
submitted in production. This is your work, and you are free to do whatever you
feel is reasonable with it. In the scenario when you don't pass, you can open
source it with any license and use it as a portfolio project.

### Areas of focus

Teleport focuses on networking, infrastructure and security.

* Reproducible pipelines. Scripts should be written in a way to allow
  reproduction of environment.
* Data models. Design performant tables. Keep queries fast and plan for schema
  migration.
* Data quality. Add validation and test coverage to all stages of your pipeline
* Security. Data should be processed and stored in a way not to leak sensitive

>   data.

Reviewer 3: Not sure of what would constitute a pass for this small project. Maybe it's just something that's mentioned in the design document that would be required for a production system.
